sklearn/cluster/dbscan_.py

```diff
@@ -147,18 +151,22 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
         labels[index] = label_num
 
         # candidates for new core samples in the cluster.
-        candidates = [index]
+        candidates = neighborhoods[index]
+        candidates = candidates[labels.take(candidates) == -1]
+
         while len(candidates) > 0:
-            # The tolist() is needed for NumPy 1.6.
-            cand_neighbors = np.concatenate(np.take(neighborhoods, candidates,
-                                                    axis=0).tolist())
-            cand_neighbors = np.unique(cand_neighbors)
-            noise = cand_neighbors[labels.take(cand_neighbors) == -1]
-            labels[noise] = label_num
-            # A candidate is a core point in the current cluster that has
-            # not yet been used to expand the current cluster.
-            candidates = np.intersect1d(noise, core_samples,
+            # Label neighbors:
+            labels[candidates] = label_num
+            # Keep only core points:
+            candidates = np.intersect1d(candidates, core_samples,
                                         assume_unique=True)
+            if len(candidates) == 0: break
+            # Get neighbors of core points:
+            for idx in np.take(neighborhoods, candidates, axis=0):
```

Comment:
Discussed somewhere before. Does not work, because it is of type list(list(int)) or something like this, depending on the index used - candidates will usually be contain more than one object, so it's the neighbors of more than one, and not a proper way of indexing an array anymore. Or so, I'm not a numpy syntax guru.